libs/common-types/src/lib.rs

```diff
@@ -130,47 +197,166 @@ where
+					PermissionedCurrencyRole::Holder(delta) => {
+						self.permissioned_asset_holder.remove(delta)
+					}
+					PermissionedCurrencyRole::Manager => {
+						Ok(self.admin.remove(AdminRoles::PERMISSIONED_ASSET_MANAGER))
+					}
+					PermissionedCurrencyRole::Issuer => {
+						Ok(self.admin.remove(AdminRoles::PERMISSIONED_ASSET_ISSUER))
+					}
+				}
+			}
 		}
 	}
 
 	fn add(&mut self, property: Self::Property) -> Result<(), ()> {
 		match property {
-			PoolRole::Borrower => Ok(self.admin.insert(AdminRoles::BORROWER)),
-			PoolRole::LiquidityAdmin => Ok(self.admin.insert(AdminRoles::LIQUIDITY_ADMIN)),
-			PoolRole::PoolAdmin => Ok(self.admin.insert(AdminRoles::POOL_ADMIN)),
-			PoolRole::PricingAdmin => Ok(self.admin.insert(AdminRoles::PRICING_ADMIN)),
-			PoolRole::MemberListAdmin => Ok(self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN)),
-			PoolRole::RiskAdmin => Ok(self.admin.insert(AdminRoles::RISK_ADMIN)),
-			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+			Role::PoolRole(pool_role) => match pool_role {
+				PoolRole::Borrower => Ok(self.admin.insert(AdminRoles::BORROWER)),
+				PoolRole::LiquidityAdmin => Ok(self.admin.insert(AdminRoles::LIQUIDITY_ADMIN)),
+				PoolRole::PoolAdmin => Ok(self.admin.insert(AdminRoles::POOL_ADMIN)),
+				PoolRole::PricingAdmin => Ok(self.admin.insert(AdminRoles::PRICING_ADMIN)),
+				PoolRole::MemberListAdmin => Ok(self.admin.insert(AdminRoles::MEMBER_LIST_ADMIN)),
+				PoolRole::RiskAdmin => Ok(self.admin.insert(AdminRoles::RISK_ADMIN)),
+				PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.insert(id, delta),
+			},
+			Role::PermissionedCurrencyRole(permissioned_currency_role) => {
+				match permissioned_currency_role {
+					PermissionedCurrencyRole::Holder(delta) => {
+						self.permissioned_asset_holder.insert(delta)
+					}
+					PermissionedCurrencyRole::Manager => {
+						Ok(self.admin.insert(AdminRoles::PERMISSIONED_ASSET_MANAGER))
+					}
+					PermissionedCurrencyRole::Issuer => {
+						Ok(self.admin.insert(AdminRoles::PERMISSIONED_ASSET_ISSUER))
+					}
+				}
+			}
+		}
+	}
+}
+
+impl<Now, MinDelay, Moment> PermissionedCurrencyHolders<Now, MinDelay, Moment>
+where
+	Now: UnixTime,
+	MinDelay: Get<Moment>,
+	Moment: From<u64> + PartialEq + PartialOrd + Saturating + Ord + Copy,
+{
+	pub fn empty() -> Self {
+		Self::default()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.info.is_none()
+	}
+
+	fn validity(&self, delta: Moment) -> Result<Moment, ()> {
+		let now: Moment = Now::now().as_secs().into();
+		let min_validity = now.saturating_add(MinDelay::get());
+		let req_validity = now.saturating_add(delta);
+
+		if req_validity < min_validity {
+			Err(())
+		} else {
+			Ok(req_validity)
+		}
+	}
+
+	pub fn contains(&self) -> bool {
+		if let Some(info) = &self.info {
+			info.permissioned_till >= Now::now().as_secs().into()
+		} else {
+			false
+		}
+	}
+
+	pub fn remove(&mut self, delta: Moment) -> Result<(), ()> {
```

Comment:
Are we sure we want the return type to be `Result<(), ()>`?

Reply:
This wasn't changed in this PR, but I think the goal is just to return whether it failed or not, there's no need to return any more information if it succeeded. What do you think should be returned otherwise?